### ISSUE TITLE
Allow for more flexibility in inheriting annotations

### ISSUE THREAD

**Issue by msimberg** [pika-org/pika#172]
Currently some annotations are inherited at the point of spawning a task. In some cases the annotations are not inherited at all (`execution::schedule/transfer`). I'm only considering sender adaptors for this and will not focus on backporting any changes decided on here to `async`/`dataflow`/`then`.

I think there are a few possibilities for determining the annotation for a task:
1. Explicitly specified annotation
2. Inheriting the annotation from the task that spawns the new task
3. Inheriting the annotation from the task that schedules the new task

By spawning I mean eagerly creating a pika thread and by scheduling I mean creating a sender that eventually creates a pika thread.

This is an example that captures the different possibilities that I can think of right now:
```
scoped_annotation ann("A"); // 1. The scheduling task has an annotation
schedule(with_annotation(thread_pool_scheduler{}, "B")) | // 2. The task is explicitly annotated "B"
  then(f1) | // 3. f1 is not annotated, do nothing or inherit "B" or "A"?
  transfer(thread_pool_scheduler{}) | // 4. Inherit "B" or "A"?
  then(annotated_function(f2, "C")) | // 5. Use "C"
  then(f3) | // 6. Inherit "C", "B", or "A"?
  start_detached();
```

In my opinion there is no discussion for 1., 2., or 5. but welcome other opinions.

In 3. I think there is no inheritance needed. The task calling f1 is already annotated with "B".

In 5. one possibility would be to change the annotation of the task to "C" so that continuations are also annotated "C". However, I think this goes against locally annotating f2, so I think "C" is not an option in 6.

That leaves 4. and whatever is the correct answer to 4. is also the answer for 6.

I can see a case for both "A" and "B" in 4. On one hand experience from DLA-Future shows that "A" is often what one would expect. On the other hand "B" is the simpler alternative (and close to the current behaviour) since one only needs to look at the context at the point of actually spawning a task. However, "simpler to implement" should be the main criterion if option "A" is more intuitive.

There is one additional situation to consider:
```
// 1. The scheduling task has no annotation or is a plain std::thread
schedule(with_annotation(thread_pool_scheduler{}, "B")) |
  then(f1) |
  transfer(thread_pool_scheduler{}) | // 2. Fall back to "B" or nothing?
  then(2) |
  start_detached();
```
If we use "inherit from the scheduling task" the label in 2. should be nothing/`<unknown>`. I think this also fits if one considers that the situation could be:
```
// 1. The scheduling task has no annotation or is a plain std::thread
some_sender_with_unknown_label_and_context |
  transfer(thread_pool_scheduler{}) | // 2. Fall back to whatever label happens to be set when transfering or nothing?
  then(2) |
  start_detached();
```

All in all, this seems to point to changing the behaviour so that annotations are by default inherited from the scheduling task, not the spawning task. If flexibility is required, we could additionally introduce options for changing how the annotations are inherited, with the default being to inherit from the scheduling task. Something like this:
```
scoped_annotation ann("A");
schedule(with_annotation(thread_pool_scheduler{}, "B")) | // Explicitly annotated with "B"
  then(f1) |
  transfer(with_annotation(thread_pool_scheduler{}, inherit_predecessor_annotation)) | // Explicitly inherit "B"
  then(f2) |
  transfer(with_annotation(thread_pool_scheduler{}, inherit_parent)) | // Inherit "A", same as default
  start_detached();
```

TL;DR:
- Function-specific annotations take precedence (this happens automatically with `annotated_function`)
- If the above is not set, the function will use an explicit annotation set on the scheduler _for one task only_
- If none of the above are set, use an annotation from the context where the task is _scheduled_ (not spawned)
- If none of the above are available, use nothing (`<unknown>` or similar), do _not_ inherit from the spawning task (unless it's the task scheduling the task as well, but this is handled by the previous point)

Additional cases to consider:
- `bulk`
- `split`/`ensure_started`
- `when_all`
- `let_x`

**Comment by rasolca** [comment]
I think that the priority for the choice should be:
1) task specific annotation
2) scheduler annotation
3) scoped annotation

e.g.
> `transfer(thread_pool_scheduler{}) | // 4. Inherit "B" or "A"?`

Should inherit "A" as the scheduler is not annotated

> `transfer(thread_pool_scheduler{}) | // 2. Fall back to "B" or nothing?`

should fall back to nothing

I hope it is clear, otherwise I can write what I expect for all the cases...

**Comment by msimberg** [comment]
Thanks @rasolca, I think that's clear and matches what makes sense to me as well.

Just to be explicit, I suppose you also don't see a need for an explicit option to in some cases inherit the annotation from the task that actually ends up spawning the new task. This is the current behaviour. That is, you don't need something like this:
```
scoped_annotation ann("A");
schedule(with_annotation(thread_pool_scheduler{}, "B")) | // Explicitly annotated with "B"
  then(f1) |
  transfer(with_annotation(thread_pool_scheduler{}, inherit_predecessor_annotation)) | // Explicitly inherit "B"
  then(f2) |
  start_detached();
```
correct? If for some reason we end up needing it in the future we can always add it, but if it's not needed right now it simplifies the implementation at least slightly.

**Comment by albestro** [comment]
Thanks @msimberg for reporting the main use-cases! I fully agree with your statements and the ones from @rasolca, I think it is a good sign that we're all expecting the same thing.

> I think that the priority for the choice should be:
>     - task specific annotation
>     - scheduler annotation
>     - scoped annotation

Yes, it's a very good schematic way to see it.
I have a similar mental model, where there is also the hierarchy of the tasks. In it, `transfer` interrupts the continuation and "re-attach" to the outer scheduling task inheriting (if there is) the annotation from it.

About this
> If flexibility is required, we could additionally introduce options for changing how the annotations are inherited, with the default being to inherit from the scheduling task.
[...]
> If for some reason we end up needing it in the future we can always add it, but if it's not needed right now it simplifies the implementation at least slightly.

IMHO we can stick to the first step of having the "fixed" policy for annotation without chances of customization. This first step will already represents a super big step forward.

**Comment by msimberg** [comment]
Considering other adaptors:

# `when_all`

Without any additional changes, the continuation of `when_all` will use the annotation of the last sender passed to `when_all` that completes. I think this is acceptable behaviour because in general it's undefined which predecessor sender will be the last to complete. `when_all` also does not advertise a `completion_scheduler` because of this. As a consequence, one should always be explicit about which scheduler to continue on after a `when_all` by doing `when_all(...) | transfer(sched)`. If one does this, the annotations will work as discussed in the previous comments. 

# `split/ensure_started`

`split`, like `when_all`, doesn't advertise a `completion_scheduler`, and where the continuation runs is undefined. Explicitly use `transfer` if you care about what annotation to have in the continuation. Otherwise you get whatever annotation was set previously.

# `bulk`

The behaviour of annotations for `bulk` with `thread_pool_scheduler` may need changing. This is somewhat dependent on the actual implementation of the `bulk` customization without any further changes to `bulk` besides the current changes in this PR.

Some examples:
```
schedule(thread_pool_scheduler{}) | // 1. inherit annotation from parent or use with_annotation
  bulk(n, [](int i) {}) | // 2. use the same annotation as 1.?
  then([]() | // 3. keep using the same annotation as 1.?
  start_detached();
```

I think the answer to 1., 2., and 3. is to use the annotation either inherited in 1. or whatever explicit annotation is set in 1. for all three steps. The reason for this is that the behaviour of `bulk` _may_ be to not spawn any extra threads. It may also be that the continuation runs on the same thread that starts `bulk`, or it may run on one of the threads that `bulk` spawns. I think the only "consistent" way of annotating would be to inherit the annotation from 1. all the way until 3. Otherwise the annotations depend on the implementation or how tasks end up running in a particular invocation of `bulk`. 

```
schedule(withthread_pool_scheduler{}) | // 1. inherit annotation from parent
  bulk(n, annotated_function([](int i) {}, "A")) | // 2. scope "A" to each invocation of the callable only, or set it for the whole parallel region?
  then([]() | // 3. use annotation from 1. or "A"?
  start_detached();
```

In this case I would in theory say that 2. should just set an annotation for each invocation of the lambda in 2. and otherwise use the annotation from 1. until 3. In practice there are overheads involved in setting annotations and setting and resetting an annotation for all calls of the lambda with `i` in `0` to `n` may be an unnecessary overhead. Because of this it may be better to set the annotation of the lambda for the whole region and then reset to the annotation from 1. after the region is done. Setting and resetting the annotation for each `i` may also be unnecessary noise. _Not_ inheriting the annotation from the `annotated_function` in 2. in step 3. would also be analogous behaviour to how `schedule(sched) | then(annotated_function(f))` behaves, i.e. `annotated_function` acts only locally.

**Comment by msimberg** [comment]
Most likely still needs some work for https://github.com/pika-org/pika/issues/172#issuecomment-1096447599.
